File: leetcode20.go
// +build ignore

package main

import "fmt"

func isValid(s string) bool {
	hash := map[rune]rune{'(': ')', '{': '}', '[': ']'}
	stack := make([]rune, len(s))
	top := -1
	for _, c := range s {
		if c == '(' || c == '{' || c == '[' {
			top++
			stack[top] = c
		} else if c == ')' || c == '}' || c == ']' {
			if top < 0 {
				return false
			}
			cpr := stack[top]
			top--
			if hash[cpr] != c {
				return false
			}
		}
	}
	if top >= 0 {
		return false
	}
	return true
}
func main() {
	fmt.Print(isValid("[]()"))
}

File: leetcode20.py
class Solution:
    def isValid(self, s: str) -> bool:
        hash = {'(':')', '{':'}', '[':']'}
        stack = []
        for char in s:
            char = str(char)
            if char == '(' or char == '{' or char == '[':
                stack.append(char)
            elif char == ')' or char == '}' or char == ']':
                if len(stack) == 0:
                    return False
                cpr = stack.pop()
                if char != hash.get(cpr):
                    return False
        if len(stack) != 0:
            return False
        return True

a = Solution()
a.isValid("[]{}")